### ISSUE TITLE
Forcefully prevent stacks from deletion

### ISSUE THREAD

**Issue by aureq** [pulumi/pulumi#9172]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

Currently, stacks can be deleted if the user has sufficient permissions, even if some resources remain in the stack using `--force` flag and even if some protected resources are still present.  This is a potential risk for customers who manage their business critical workloads via Pulumi as they don't want to loose the state of their stack.

Similarly to the EC2 termination protection, some possible enhancements:

1 - A stack cannot be deleted if a `protected` resource remains present in that stack. Maybe a feature flag to be enabled/disabled at an account or org level could be helpful as a way to allow stack deletion when something is broken in the stack code.
2 - A stack cannot be deleted if `stack termination protection` is enabled (Updates should be allowed). The toggle for this feature could require a separate CLI command to avoid bundling options like `--force` on a single command. On `pulumi up`, a warning could be displayed to signal this stack is sensitive/protected.

### Affected area/feature

* Pulumi Service (Web Console)
* Pulumi CLI

**Comment by Frassle** [comment]
Feel like maybe this should be a service feature?

**Comment by aureq** [comment]
Yes, partly.

In case of 1, then this could simply be enforced within the CLI regardless of the backend used. Maybe by checking any resources with the protected flag in the stack state file.

 As for 2, this would definitely be a pulumi service feature.

**Comment by Frassle** [comment]
OK so part 1 is adding an extra option so that --force alone will only delete the stack if it's got non-protected resources in it.
So we get:
1) No options, can't delete a stack with resources in it
2) --force, can't delete a stack with protected resources in it, but can if they're just un-protected resources
3) --XYZ, can delete a stack regardless of what's in it

Any suggestions for that flag name? `--include-protected`?

**Comment by aureq** [comment]
Yes you're right @Frassle 👍 Though, the `--force` wouldn't work if the stack is connected to the Pulumi service and stack termination is enable for that stack.

A possible flag name could be `--delete-protected` ? 

Also, would you like me to open a separate issue in the Pulumi service repo to cover the 2nd point ?

**Comment by Frassle** [comment]
> --delete-protected ?

Hmm delete protected suggests its actually doing deletes, and we've already got an issue about this command having confusing wording https://github.com/pulumi/pulumi/issues/9166

> Also, would you like me to open a separate issue in the Pulumi service repo to cover the 2nd point ?

https://github.com/pulumi/service-requests/issues please 😄

**Comment by aureq** [comment]
🤔... yes you have an excellent point ! finding a good flag it tricky 🤨

How about `--ignore-protected-resources` or maybe `--ignore-stack-protected-resources`. Then the command would be `pulumi stack rm <stack_name> --force --ignore-protected-resources` ?

It feels we should be explicitly using `resources` in the flag name so it's clear enough. Other than `ignore`, we could also use `bypass` or `override`.

**Comment by Frassle** [comment]
Had a look at the code needed for this. We'd need a service change for the first part as well, currently the CLI just sends a DELETE request to the service with a "force" flag and then the service checks if there's any resources in the stack. So we need the service to support "force" and "something about protect" on the stack DELETE handler before we add it to the CLI.
